src/hdmf/container.py

```diff
@@ -747,22 +747,34 @@ def __smart_str_dict(d, num_indent):
         out += '\n' + indent + right_br
         return out
 
-    def set_data_io(self, dataset_name: str, data_io_class: Type[DataIO], **kwargs) -> None:
+    def set_data_io(self, dataset_name: str, data_io_class: Type[DataIO], data_io_kwargs: dict = None, **kwargs):
         """
+        Apply DataIO object to a dataset field of the Container.
 
         Parameters
         ----------
         dataset_name: str
             Name of dataset to wrap in DataIO
         data_io_class: Type[DataIO]
             Class to use for DataIO, e.g. H5DataIO or ZarrDataIO
+        data_io_kwargs: dict, optional
+            keyword arguments passed to the constructor of the DataIO class.
         **kwargs:
-            kwargs are passed to the constructor of the DataIO class
+            DEPRECATED. Use data_io_kwargs instead.
+            kwargs are passed to the constructor of the DataIO class.
         """
+        if kwargs or (data_io_kwargs is None):
```

Comment:
The warning triggers when `data_io_kwargs` is not provided, which I think is not intended.
However, I can't imagine why someone would wrap a dataset with a `DataIO` without any kwargs. So I suggest that `data_io_kwargs` is required.

Reply:
However, that would make a hard breaking change, because `obj.set_data_io("data2", H5DataIO, chunks=True)` would fail, so never mind.

Reply:
Yes, that's the dilemma.